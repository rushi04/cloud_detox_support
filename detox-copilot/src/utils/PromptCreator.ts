export class PromptCreator {
    constructor(private availableAPI: TestingFrameworkAPI) {}

    createPrompt(
        step: ExecutionStep,
        viewHierarchy: string,
        isSnapshotImageAttached: boolean,
        previousSteps: ExecutionStep[]
    ): string {
        return [
            this.createBasePrompt(),
            this.createContext(step, viewHierarchy, isSnapshotImageAttached, previousSteps),
            this.createAPIInfo(step.type),
            this.createInstructions(step, isSnapshotImageAttached)
        ]
            .flat()
            .join('\n');
    }

    private createBasePrompt(): string[] {
        return [
            "# Test Code Generation",
            "",
            "You are an AI assistant tasked with generating test code for a mobile application using the provided UI testing framework API.",
            "Please generate the minimal executable code to perform the desired intent based on the given information and context.",
            ""
        ];
    }

    private createContext(
        step: ExecutionStep,
        viewHierarchy: string,
        isSnapshotImageAttached: boolean,
        previousSteps: ExecutionStep[]
    ): string[] {
        let context = [
            "## Context",
            "",
            "### Step to perform",
            "",
            `Generate the minimal executable code to perform the following ${step.type}: "${step.value}"`,
            "",
            "### View hierarchy",
            "",
            "```",
            `${viewHierarchy}`,
            "```",
            ""
        ];

        if (isSnapshotImageAttached) {
            context.push(
                "### Snapshot image",
                "",
                "Snapshot image is attached for visual reference.",
                ""
            );
        }

        if (previousSteps.length > 0) {
            context.push(
                "### Previous steps",
                "",
                ...previousSteps.map((prevStep, index) => `${index + 1}. ${prevStep.type}: ${prevStep.value}`),
                ""
            );
        }

        return context;
    }

    private createAPIInfo(stepType: ExecutionStepType): string[] {
        const relevantAPI = stepType === 'action' ? this.availableAPI.actions : this.availableAPI.assertions;
        return [
            `## Available ${stepType.charAt(0).toUpperCase() + stepType.slice(1)} API`,
            ""
        ]
            .concat(
                relevantAPI.map(this.formatAPIMethod).flat()
            );
    }

    private formatAPIMethod(method: TestingFrameworkAPIMethod): string[] {
        const methodInfo = [
            `### ${method.signature}`,
            "",
            method.description,
            "",
            "#### Example",
            "",
            "```",
            method.example,
            "```",
            ""
        ];

        if (method.guidelines.length > 0) {
            methodInfo.push(
                "#### Guidelines",
                ...method.guidelines.map(guideline => `- ${guideline}`)
            );
        }

        return methodInfo;
    }

    private createInstructions(step: ExecutionStep, isSnapshotImageAttached: boolean): string[] {
        return [
            "## Instructions",
            "",
            [
                `Generate the minimal executable code to perform the following ${step.type}: "${step.value}"`,
                "Use the provided API and follow the guidelines.",
                "If you cannot generate the relevant code due to ambiguity, invalid prompt, or inability to find the desired element, return a code that throws an informative error explaining the problem in one sentence.",
                "Wrap the generated code with backticks, without any additional formatting.",
            ]
                .concat(this.createVisualAssertionsInstructionIfPossible(isSnapshotImageAttached))
                .map((instruction, index) => `${index + 1}. ${instruction}`).join('\n'),
            "",
            "### Examples of throwing an informative error:",
            "```typescript",
            'throw new Error("Unable to find the \'Submit\' button in the current view hierarchy.");',
            "```",
            "",
            "```typescript",
            'throw new Error("The provided prompt does not contain enough information to generate the code, \"button\" is too ambiguous for matching a specific element.");',
            "```",
            "",
            "Please provide your response below:"
        ];
    }

    private createVisualAssertionsInstructionIfPossible(isSnapshotImageAttached: boolean): string[] {
        if (!isSnapshotImageAttached) {
            return [];
        }

        return [
            "In case the expected behaviour can be tested visually based on the provided snapshot image, there's no need to generate test code for the assertion. " +
                "Instead, return code that throws an error if the visual check fails, or an empty code block if the visual check passes.",
        ];
    }
}
