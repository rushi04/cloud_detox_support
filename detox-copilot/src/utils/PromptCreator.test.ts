import { PromptCreator } from './PromptCreator';

const mockAPI: TestingFrameworkAPI = {
    actions: [
        {
            signature: 'tap(element: Element)',
            description: 'Taps on the specified element.',
            example: 'await element(by.id("button")).tap();',
            guidelines: ['Ensure the element is tappable before using this method.']
        },
        {
            signature: 'typeText(element: Element, text: string)',
            description: 'Types the specified text into the element.',
            example: 'await element(by.id("input")).typeText("Hello, World!");',
            guidelines: ['Use this method only on text input elements.']
        }
    ],
    assertions: [
        {
            signature: 'toBeVisible()',
            description: 'Asserts that the element is visible on the screen.',
            example: 'await expect(element(by.id("title"))).toBeVisible();',
            guidelines: ['Consider scroll position when using this assertion.']
        }
    ],
    matchers: [
        {
            signature: 'by.id(id: string)',
            description: 'Matches elements by their ID attribute.',
            example: 'element(by.id("uniqueId"))',
            guidelines: ['Use unique IDs for elements to avoid conflicts, combine with atIndex() if necessary.']
        }
    ]
};

describe('prompt creation', () => {
    let promptCreator: PromptCreator;

    beforeEach(() => {
        promptCreator = new PromptCreator(mockAPI);
    });

    it('should create an action step prompt correctly', () => {
        const step: ExecutionStep = {
            type: 'action',
            value: 'tap button'
        };

        const viewHierarchy = '<View><Button testID="submit" title="Submit" /></View>';
        const prompt = promptCreator.createPrompt(step, viewHierarchy, true, []);
        expect(prompt).toMatchSnapshot();
    });

    it('should create an assertion step with snapshot image prompt correctly', () => {
        const step: ExecutionStep = {
            type: 'assertion',
            value: 'expect button to be visible'
        };

        const viewHierarchy = '<View><Button testID="submit" title="Submit" /></View>';
        const prompt = promptCreator.createPrompt(step, viewHierarchy, true, []);
        expect(prompt).toMatchSnapshot();
    });

    it('should create an assertion step without snapshot image prompt correctly', () => {
        const step: ExecutionStep = {
            type: 'assertion',
            value: 'expect button to be visible'
        };

        const viewHierarchy = '<View><Button testID="submit" title="Submit" /></View>';
        const prompt = promptCreator.createPrompt(step, viewHierarchy, false, []);
        expect(prompt).toMatchSnapshot();
    });
});
