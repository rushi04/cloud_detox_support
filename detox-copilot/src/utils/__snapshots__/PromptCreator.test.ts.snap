// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prompt creation should create an action step prompt correctly 1`] = `
"# Test Code Generation

You are an AI assistant tasked with generating test code for a mobile application using the provided UI testing framework API.
Please generate the minimal executable code to perform the desired intent based on the given information and context.

## Context

### Step to perform

Generate the minimal executable code to perform the following action: "tap button"

### View hierarchy

\`\`\`
<View><Button testID="submit" title="Submit" /></View>
\`\`\`

### Snapshot image

Snapshot image is attached for visual reference.

## Available Action API

### tap(element: Element)

Taps on the specified element.

#### Example

\`\`\`
await element(by.id("button")).tap();
\`\`\`

#### Guidelines
- Ensure the element is tappable before using this method.
### typeText(element: Element, text: string)

Types the specified text into the element.

#### Example

\`\`\`
await element(by.id("input")).typeText("Hello, World!");
\`\`\`

#### Guidelines
- Use this method only on text input elements.
## Instructions

1. Generate the minimal executable code to perform the following action: "tap button"
2. Use the provided API and follow the guidelines.
3. If you cannot generate the relevant code due to ambiguity, invalid prompt, or inability to find the desired element, return a code that throws an informative error explaining the problem in one sentence.
4. Wrap the generated code with backticks, without any additional formatting.
5. In case the expected behaviour can be tested visually based on the provided snapshot image, there's no need to generate test code for the assertion. Instead, return code that throws an error if the visual check fails, or an empty code block if the visual check passes.

### Examples of throwing an informative error:
\`\`\`typescript
throw new Error("Unable to find the 'Submit' button in the current view hierarchy.");
\`\`\`

\`\`\`typescript
throw new Error("The provided prompt does not contain enough information to generate the code, "button" is too ambiguous for matching a specific element.");
\`\`\`

Please provide your response below:"
`;

exports[`prompt creation should create an assertion step with snapshot image prompt correctly 1`] = `
"# Test Code Generation

You are an AI assistant tasked with generating test code for a mobile application using the provided UI testing framework API.
Please generate the minimal executable code to perform the desired intent based on the given information and context.

## Context

### Step to perform

Generate the minimal executable code to perform the following assertion: "expect button to be visible"

### View hierarchy

\`\`\`
<View><Button testID="submit" title="Submit" /></View>
\`\`\`

### Snapshot image

Snapshot image is attached for visual reference.

## Available Assertion API

### toBeVisible()

Asserts that the element is visible on the screen.

#### Example

\`\`\`
await expect(element(by.id("title"))).toBeVisible();
\`\`\`

#### Guidelines
- Consider scroll position when using this assertion.
## Instructions

1. Generate the minimal executable code to perform the following assertion: "expect button to be visible"
2. Use the provided API and follow the guidelines.
3. If you cannot generate the relevant code due to ambiguity, invalid prompt, or inability to find the desired element, return a code that throws an informative error explaining the problem in one sentence.
4. Wrap the generated code with backticks, without any additional formatting.
5. In case the expected behaviour can be tested visually based on the provided snapshot image, there's no need to generate test code for the assertion. Instead, return code that throws an error if the visual check fails, or an empty code block if the visual check passes.

### Examples of throwing an informative error:
\`\`\`typescript
throw new Error("Unable to find the 'Submit' button in the current view hierarchy.");
\`\`\`

\`\`\`typescript
throw new Error("The provided prompt does not contain enough information to generate the code, "button" is too ambiguous for matching a specific element.");
\`\`\`

Please provide your response below:"
`;

exports[`prompt creation should create an assertion step without snapshot image prompt correctly 1`] = `
"# Test Code Generation

You are an AI assistant tasked with generating test code for a mobile application using the provided UI testing framework API.
Please generate the minimal executable code to perform the desired intent based on the given information and context.

## Context

### Step to perform

Generate the minimal executable code to perform the following assertion: "expect button to be visible"

### View hierarchy

\`\`\`
<View><Button testID="submit" title="Submit" /></View>
\`\`\`

## Available Assertion API

### toBeVisible()

Asserts that the element is visible on the screen.

#### Example

\`\`\`
await expect(element(by.id("title"))).toBeVisible();
\`\`\`

#### Guidelines
- Consider scroll position when using this assertion.
## Instructions

1. Generate the minimal executable code to perform the following assertion: "expect button to be visible"
2. Use the provided API and follow the guidelines.
3. If you cannot generate the relevant code due to ambiguity, invalid prompt, or inability to find the desired element, return a code that throws an informative error explaining the problem in one sentence.
4. Wrap the generated code with backticks, without any additional formatting.

### Examples of throwing an informative error:
\`\`\`typescript
throw new Error("Unable to find the 'Submit' button in the current view hierarchy.");
\`\`\`

\`\`\`typescript
throw new Error("The provided prompt does not contain enough information to generate the code, "button" is too ambiguous for matching a specific element.");
\`\`\`

Please provide your response below:"
`;
